File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/NewmanPrime.c
#include "java_arrays.h"
#include <stdint.h>

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left - right);
*/
static int32_t java_sub(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left - (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left * right);
*/
static int32_t java_mul(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left * (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
  logic integer fb_value(integer n) = n <= 1 ? 1 : 2 * fb_value(n - 1) + fb_value(n - 2);
*/
/*@
  requires n >= 0;
  decreases n;
  assigns \nothing;
  ensures \result == (int32_t)fb_value(n);
*/
int32_t newmanPrime(int32_t n) {
    if (n == 0 || n == 1) {
        return 1;
    }
    int32_t first = newmanPrime(java_sub(n, 1));
    int32_t second = newmanPrime(java_sub(n, 2));
    return java_add(java_mul(2, first), second);
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/NewmanPrime.java
import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;


class NewmanPrime {
    
    /*@ public normal_behavior
      @ ensures \result == ((x % 4294967296L + 6442450944L) % 4294967296L) - 2147483648L;
      @ model public static pure \bigint fb_wrap(\bigint x);
      @*/
    /*@ public normal_behavior
      @ requires n >= 0;
      @ ensures \result == (n <= 1 ? 1 : 2 * fb_value(n - 1) + fb_value(n - 2));
      @ measured_by n;
      @ model public static pure \bigint fb_value(\bigint n);
      @*/
    /*@ public normal_behavior
      @ requires n >= 0;
      @ assignable \nothing;
      @ ensures \result == fb_wrap(fb_value(n));
      @ measured_by n;
      @*/
    //@ code_java_math
    public static int newmanPrime(int n) {
        if (n == 0 || n == 1) {
            return 1;
        }
        return 2 * newmanPrime(n - 1) + newmanPrime(n - 2);
    }
}

